frontend/model/contracts/group.js

```diff
@@ -259,6 +259,14 @@ DefineContract({
         }
       }
     },
+    'gi.contracts/group/updateSettings': {
+      validate: object,
```

Comment:
Per [this other comment](https://github.com/okTurtles/group-income-simple/pull/679/files#r328900176), here we need to validate at least: `mincomeAmount` and `mincomeCurrency`.

We could do that by doing:

```js
validate: objectOf({
  mincomeAmount: number,
  mincomeCurrency: string
})
```

However, this action will then throw if any other settings are set. Also, we would want to do additional validation to make sure that `mincomeAmount` is `> 0`.

So what we want is a validation function here that throws a `new TypeError` if any of the keys *we know we might get* do not pass our validations, and doesn't throw if we receive unknown keys.

To do this you will need to write a custom function (and make sure to thoroughly test it to verify that it satisfies the above properties).

Something like:

```js
validate: o => {
  const validations = {
    mincomeAmount: x => typeof x === 'number' && x > 0,
    mincomeCurrency: x => typeof x === 'string'
  }
  for (const key in o) {
    if (validations[key]) {
      if (!validations[key](o[key])) {
        throw new TypeError(`${key} has bad value: ${o[key]}`)
      }
    } else {
      // ignore keys for which we have no validation defined
    }
  }
  return o
}
```

(Just test and verify that works, I wrote this without testing)

Reply:
Note that `TypeError` is specifically chosen to trigger [`GIErrorIgnoreAndBanIfGroup`](https://github.com/okTurtles/group-income-simple/blob/master/frontend/model/state.js#L412-L418) if it fails. Alternatively we could throw `GIErrorIgnoreAndBanIfGroup` directly here.

Reply:
I did as you suggested. However, I think this validation part can get really verbose as we go, so we should work later on a way to optimize it...

I have a question regarding the usage of Flow. as far as I understand flow is a type checker for development only. Once in production, all flow checkers are removed. That way, how do we make sure the data is still valid?

Reply:
> I have a question regarding the usage of Flow. as far as I understand flow is a type checker for development only. Once in production, all flow checkers are removed. That way, how do we make sure the data is still valid?

These validations are what guarantee that.

These are runtime validations. We are using a runtime type checking system modeled after flow types, in addition to the dev-time static type checker.